### ISSUE TITLE
Support for ChaCha20 testvectors?

### ISSUE THREAD

**Issue by woodruffw** [google/wycheproof#90]
Hi there!

I'm opening this issue to see if there's appetite for supporting ChaCha20 test vectors (i.e. just the block cipher, not a composed AEAD).

As justification:

* ChaCha20 is specified in [RFC 7539](https://datatracker.ietf.org/doc/html/rfc7539) with a 96/32 nonce/counter split;
* ChaCha20 is also originally specified [here](https://cr.yp.to/chacha/chacha-20080128.pdf), and doesn't directly specify a nonce/counter split; a common split is 64/64 nonce/counter.

The variance between the two has produced some discrepancies between major implementations of ChaCha20. From a survey:

* OpenSSL takes a 32-bit counter as input, but rolls over to the next 32 bits (in the nonce space) on counter overflow. [Source](https://github.com/openssl/openssl/blob/b134300a342476398c11c19af602d7b2aa6b7f8a/crypto/evp/e_chacha20_poly1305.c#L112-L113).
* LibreSSL takes a 64-bit counter as input, and increments the high 32 bits on overflow of the lower bits. [Source](https://github.com/libressl/openbsd/blob/58f1458d9d651a74790afcaf4d0ff2ee2abed265/src/lib/libcrypto/chacha/chacha-merged.c#L236-L238)
* BoringSSL takes a 32-bit counter as input, and only uses those 32-bits. On overflow the counter wraps around. [Source](https://github.com/google/boringssl/blob/b0a026f8541c551854efd617021bb276f1fe5c23/crypto/chacha/chacha.c#L169)

Of the three, BoringSSL's implementation is the only one that strictly follows the RFC. 

Naively, this should only cause issues at the the 2^32 block boundary, assuming that the counter starts at 0 or 1. However, starting at 0 or 1 is not standardized, and some ChaCha20 APIs assume that the entire 16-byte nonce+counter input space is randomly initialized instead. This means that the initial counter value _may_ be significantly closer to the rollover point than 2^32 blocks, causing the different implementations to diverge after fewer enciphered blocks than might otherwise be expected.

CC @davidben, who I've been nagging about this 🙂

**Comment by davidben** [comment]
I think the right way to think about this is that IETF ChaCha20 and 64/64 ChaCha20 are related but different primitives. It is unfortunate that they have the same name, but ideally the world would settle on the IETF one as that's the point of standards. 

> OpenSSL takes a 32-bit counter as input, but rolls over to the next 32 bits (in the nonce space) on counter overflow. [Source](https://github.com/openssl/openssl/blob/b134300a342476398c11c19af602d7b2aa6b7f8a/crypto/evp/e_chacha20_poly1305.c#L112-L113).

Where are you getting the 32-bit input? That doesn't seem right. OpenSSL seems just believe in the 64/64 split. It takes as input a combined 128-bit nonce+counter (`CHACHA_CTR_SIZE` is 16 bytes) and, when incrementing, increments the bottom 64 bits.

That's perfectly coherent. It's just not the same primitive as the IETF 96/32 version. Though their documentation references the IETF version, so there's a documentation error there. I'll file a bug about that.
https://www.openssl.org/docs/manmaster/man3/EVP_chacha20_poly1305.html

**Comment by davidben** [comment]
Also keep in mind that wanting to start the block counter at a high value like you're suggesting doesn't make much sense. So while they are formally different primitives, if your application especially cares about the difference, it's probably doing something wrong, or at least very very obscure[*]. :-)

[*] QUIC packet number protection does pass in an arbitrary high value, but that's because it doesn't want ChaCha20 the stream cipher. It wants ChaCha20 the block function. There's no actual incrementing going on.

**Comment by davidben** [comment]
There _can_ be a reason to specify the counter if you're trying to reserve some counter values for miscellaneous things (as the AEAD does), or resuming a stream across two calls. But those won't get you large values out of thin air.

**Comment by davidben** [comment]
Arguably the spec doesn't actually say you're supposed to wraparound. Though it mostly doesn't say anything either way:

> ChaCha20 successively calls the ChaCha20 block function, with the
> same key and nonce, and with successively increasing block counter
> parameters. ChaCha20 then serializes the resulting state by writing
> the numbers in little-endian order, creating a keystream block.

https://www.rfc-editor.org/rfc/rfc7539.html#section-2.4

Wrapping isn't "successively increasing". But it also doesn't say what to do if you can't increase. Then we have...

> `key_stream = chacha20_block(key, counter+j, nonce)`

https://www.rfc-editor.org/rfc/rfc7539.html#section-2.4.1

Who knows what the normative semantics of that pseudocode is. :-)

**Comment by woodruffw** [comment]
Yeah, the only thing I see in the spec that implies wraparound is about the round structure:

>    Note: "addition" in the above paragraph is done modulo 2^32.  In some
   machine languages, this is called carryless addition on a 32-bit
   word.

https://www.rfc-editor.org/rfc/rfc7539.html#section-2.3

But that doesn't say anything about wraparound in the block counter itself.

I'm not too familiar with the processes here -- do you think this is worthy of an errata? The RFC is not prescriptive about the counter's initial value (it suggests 0 or 1, but only as a suggestion), so emphasizing that the counter's increment is defined modulo 2^32 would eliminate at least one point of ambiguity.

**Comment by davidben** [comment]
> The RFC is not prescriptive about the counter's initial value (it suggests 0 or 1, but only as a suggestion), so emphasizing that the counter's increment is defined modulo 2^32 would eliminate at least one point of ambiguity.

I think the RFC is actually decently clear about the initial value. It says:

>   o  A 32-bit initial counter.  This can be set to any number, but will
>      usually be zero or one.  It makes sense to use one if we use the
>      zero block for something else, such as generating a one-time
>      authenticator key as part of an AEAD algorithm.

It can't _actually_ prescribe zero or one, because, say, five would be perfectly reasonable if you need to reserve blocks 0-4 for something.

Given the mess here, I think it's pretty clear that, whether the primitive is defined to wrap or not, you really should not rely on it. So, one way or another, the spec shouldn't emphasize a modulo increment and instead should emphasize that you should avoid wrapping.

Really the problem is ChaCha20 is a low-level building block that you use to build _actual_ primitives like AEADs. It's not something you should use if you aren't willing to think about your nonce and counter space and whether you're trying to partition it. We definitely should nail down the semantics, but ultimately it sounds like pyca/cryptography has exposed it wrong.

**Comment by woodruffw** [comment]
Revising this: given that the IETF and "original"/SSH variants are both widely implemented, my current thinking is that it probably makes sense to have separate Wycheproof vectors for both.

Does that seem reasonable?

**Comment by bleichenbacher-daniel** [comment]
I looked into SSH recently. Since the maximal amount of data that can be sent before rekeying is 1 or 4 GB a counter will never overflow a 32 bit boundary. Hence implementations can use a standard implementation without problem. 

There may be some value in generating test vectors for SSH itself. I have some initial code for testing certificates. However, I can only work in my free time hence I can not make any promises.   I'm no longer working at Google and don't have access to documents and generation code anymore. This means that a lot of my free time goes into reimplementing stuff from scratch.

**Comment by bleichenbacher-daniel** [comment]
Just some related stuff.
NaCl uses some ChaCha/Salsa variants that are different from the IETF versions. In particular, there is no support for additional data.
The new test vectors can be found here:
https://github.com/bleichenbacher-daniel/Rooterberg/tree/main/test_vectors/auth_enc
